### ISSUE TITLE
How to decode the result of convictionVotes

### ISSUE THREAD

**Issue by Tbaut** [polkadot-api/polkadot-api#666]
Hello there, I'm motivated tonight :)
I'm hitting a roadblock with the result and the type of `api.query.ConvictionVoting.VotingFor`
Here is a piece of code that returns the following
```
    api.query.ConvictionVoting.VotingFor.getValue(
      "1ThiBx5DDxFhoD9GY6tz5Fp4Y7Xn1xfLmDddcoFQghDvvjg",
      0,
    ).then((res) => {
      console.log('------->res', res)
    })
```
![image](https://github.com/user-attachments/assets/a859b33c-8ed9-4c5f-8a55-d02eea7847a4)

The type is `ConvictionVotingVoteAccountVote`, but for the "Standard" entry, the vote is then a number (it's 3 in this specific case).

```
export type ConvictionVotingVoteAccountVote = Enum<{
    "Standard": Anonymize<Ib024p97ls1cla>;
    "Split": Anonymize<I5pi71t9bosoiv>;
    "SplitAbstain": Anonymize<I89irppcaqmf1i>;
}>;
export declare const ConvictionVotingVoteAccountVote: GetEnum<ConvictionVotingVoteAccountVote>;
export type Ib024p97ls1cla = {
    "vote": number;
    "balance": bigint;
};
```

However, I'd expect it to be something like `vote: {conviction: ConvictionVoting; amount: VoteAyeNay}` as the result in pjs suggests:

```
    {
      Casting: {
        votes: [
          [
            785
            {
              Standard: {
                vote: {
                  conviction: Locked3x  <--------------------------------------
                  vote: Nay  <--------------------------------------
                }
                balance: 50,000,000,000
              }
            }
          ]
        ]
        delegations: {
          votes: 0
          capital: 0
        }
        prior: [
          0
          0
        ]
      }
    }
```

How do I go from `3` to `{ conviction: "Locked3x", vote: "Nay"}`?

**Comment by josepot** [comment]
Yes, we are well aware of this particular issue.

The problem here is that according to the metadata the vote is a `u8`, so the coded and the typings treat it as a number... That's because in the pallet is [being encoded like this](https://github.com/paritytech/polkadot-sdk/blob/efdc1e9b1615c5502ed63ffc9683d99af6397263/substrate/frame/conviction-voting/src/vote.rs#L36-L53) 😞 

If you ask me I would say that's a total overkill. Sure, one byte less for each vote reduces the fees, but is it really worth it taking into account the complexity that it adds on the client side? I don't think so.

Anyways, unless the Metadata V16 finds a way to address this (which I don't think it's going to be the case). The only solution to this problem is either:
1) Have pallet-oriented SDKs that have custom logic for each pallet and tackle this issue (what we are currently working on).
2) Ask users to use functions that can re-decode these kinds of values. E.g:

```ts
interface Vote {
  aye: boolean
  conviction: number
}

const getVoteFromNumber = (input: number): Vote => ({
  aye: Boolean(input & 0b1000_0000),
  conviction: input & 0b0111_1111,
})

const getNumberFromVote = ({ aye, conviction }: Vote): number =>
  +aye * 0b1000_0000 + conviction
```
I have to say that this case is a very extreme case and that I'm not aware of any other cases as extreme as this one. That being said, this is something that IMO should definitely be improved on PolkadotSDK, either by improving the metadata or by "wasting" one more byte.

**Comment by Tbaut** [comment]
Thanks a lot for the explanation and the helper functions. I imagine that this will eventually be part of papi as a helper, or that it will be done transparently and decided nicely eventually? I think it already is done for transaction, as I remember submitting a conviction vote with a string and it got encoded correctly.

**Comment by bkchr** [comment]
> That being said, this is something that IMO should definitely be improved on PolkadotSDK, either by improving the metadata or by "wasting" one more byte.

I get your pain. However, you need to see this type as any other basic type which also have their own encoding.

**Comment by josepot** [comment]
> However, you need to see this type as any other basic type which also have their own encoding.

In the short term, treating it this way might work - as we are already doing with the pallet-based SDKs. However, the metadata should ideally be self-descriptive. Currently, there isn't a straightforward way to achieve what you're suggesting without resorting to a workaround, and such an approach doesn't scale well.

For example, when a user signs a transaction on a device like a Ledger or any other device that relies on metadata to display transaction details, the `vote` will appear as a simple `number` (`u8`). Unfortunately, there's no automatic way to provide a more meaningful display based solely on the metadata. This limitation is a significant drawback.

One potential solution is to encode the vote as an enum, representing all possible permutations (e.g., `AyeNoConviction`, `AyeLocked1`, `AyeLocked2`, `NayNoConviction`, etc.). This approach would still only require one byte (which would map all options to the same value) but would allow for better type definitions and provide users with more informative displays.

I may work on a PR to PolkadotSDK this week to implement this.

Alternatively, there should be a mechanism to describe this type of "bit boxing" directly in the metadata.

**Comment by kianenigma** [comment]
What are some examples of these pallet-based SDKs? 

Perhaps we can move as much of them as possible to the runtimes, as pallet-specific runtime APIs. The goal should be to have TS libraries not need to rewrite arbitary Rust logic on their side. Something like governance lock details can probably be easily solved in the runtime API level better. wdyt?

**Comment by bkchr** [comment]
> The goal should be to have TS libraries not need to rewrite arbitary Rust logic on their side. Something like governance lock details can probably be easily solved in the runtime API level better. wdyt?

This is like saying the TS side should not implement SCALE decoding of `u32`, because it is a re-implementation of the Rust logic.

> For example, when a user signs a transaction on a device like a Ledger or any other device that relies on metadata to display transaction details, the `vote` will appear as a simple `number` (`u8`).

They already don't know what a Balance is and what is some other random value. 

We need semantic information and this could maybe help here as well.

**Comment by josepot** [comment]
First and foremost, I want to express my deep appreciation for the fact that we have 2 core contributors of polkadot caring about this 🙏. It's truly remarkable to see how much you care about this issue, so thank you.

I'm currently considering making an exception in this particular case by "hacking" the type generation process to treat the conviction vote as a top-level primitive. What I mean by this is that, for certain special cases, we use custom logic based on heuristics to detect specific types that we consider "first-class citizens" in the system. These include types like addresses (both `AccountId20` and `AccountId32`), binary data (such as `u8` arrays), `Option`s, and `Result`s. While the custom logic we employ to detect these types from the metadata isn't fully documented, it's generally reliable and low-risk in my opinion. However, we aim to minimize the use of heuristic-based logic wherever possible, limiting it only to very common types.

Introducing custom logic to detect a type like the "vote" in the referenda feels like crossing a line I initially wanted to avoid. Nevertheless, given that this type is from an older and widely used pallet, it might be pragmatic to introduce a small piece of custom logic to handle its detection during the generic type generation process. We're still debating this within the team, but it might be the most practical solution at this point.

That said, it’s important that, moving forward, we strive to avoid relying on custom logic on the client side as much as possible. As I mentioned earlier, metadata should ideally be self-descriptive. This means that either pallets should avoid using bit-boxing techniques, or we need to find a way to communicate these details clearly within the metadata itself.

---------------------

> What are some examples of these pallet-based SDKs?

From our latest OpenGov proposal:

> When creating `polkadot-api`, we strictly focused on building a chain-agnostic library with no business model logic. While low-level endpoints are useful for advanced implementers, we need higher-level APIs that understand the pallet's business logic. The main challenge we've identified in Polkadot is not the UI components but the highly low-level nature of the blockchain's APIs. End users should not need to worry about storage queries, runtime calls, or other low-level implementation details of Polkadot, yet they currently do.

The core idea is that the current polkadot-api provides excellent TypeScript bindings based on the metadata, along with useful utilities for cross-cutting concerns like generic fee calculation, crafting transactions, etc.

Our next focus is on developing libraries that encapsulate key business logic, so DApp developers don't have to reinvent the wheel every time. For example, we could create an SDK specifically for handling delegation. This SDK would manage the logic required for delegation, such as determining the necessary actions, considering side-effects (e.g., has the user voted recently? Are they already delegating? Are there any [potential risks](https://wiki.polkadot.network/docs/maintain/maintain-guides-democracy#undelegate-a-vote) the user should be aware of before undelegating?), and generating an optimal batch call, plus the necessary information to be communicated to the user.

In essence, it's about demonstrating how you can build on what polkadot-api offers OOTB to create meaningful, reusable business logic for DApps.

**Comment by kianenigma** [comment]
> > The goal should be to have TS libraries not need to rewrite arbitary Rust logic on their side. Something like governance lock details can probably be easily solved in the runtime API level better. wdyt?
> 
> This is like saying the TS side should not implement SCALE decoding of `u32`, because it is a re-implementation of the Rust logic.

SCALE is so fundamental that it has to be. But do you say that TS libraries should all re-implement things like:

1. how much is the actual inflation rate?
2. how much of a user's balance is transferrable 

and on son? I hope the answer is no :) 

ofc this can still be in a shared TS library, but historically we've seen that all wallets implementing their own version is bad. 

Even to this day, if I go to 3 different UIs, I get 3 different answer to the above two queries. There's obviously a need for standardization here.

**Comment by Tbaut** [comment]
> how much of a user's balance is transferable

I thought about this a couple days, ago, but given that the answer is [in the wiki](https://wiki.polkadot.network/docs/learn-guides-accounts#query-account-data-in-polkadot-js) (somewhat buried).. I didn't mention it. But you're right, this is such an important value, it should be available to Dapps with 1 query and no calculation.

**Comment by kianenigma** [comment]
> > how much of a user's balance is transferable
> 
> I thought about this a couple days, ago, but given that the answer is [in the wiki](https://wiki.polkadot.network/docs/learn-guides-accounts#query-account-data-in-polkadot-js) (somewhat buried).. I didn't mention it. But you're right, this is such an important value, it should be available to Dapps with 1 query and no calculation.

Exactly, requiring/expecting every DApp developer to re-invent this wheel, _even if it is done correctly_, seems wasteful to me. But as noted, my personal anecdote is that it is not done correctly. When I log in to PJS apps and Talisman, I get two very different values as to "how much liquid DOT I have in total". Is it just me?

**Comment by bkchr** [comment]
> Even to this day, if I go to 3 different UIs, I get 3 different answer to the above two queries. There's obviously a need for standardization here.

Yeah for sure. This was the idea of having models that can be used by anyone to handle these kind of tasks and having a stable interface. In the end it requires some thinking on where you want to do this. If you want to use some runtime apis for this job or we write something that works outside of the runtime. On the first glance using runtime apis for this sounds natural and also easy. However, if you think more about this, it also means that you put all these computations on the nodes. Then you will need to redo the computations on the light client. For sure individual requests will not be a problem, but the more users you get, the more requests you get etc. Just pushing out the data to the user and let them do the computations locally should be more efficient.

**Comment by kianenigma** [comment]
@josepot on pallet/business-logic level APIs::

Direction: 100% agree. 

Implementation: It is, end of the day, up to you and your maintenance cost to consider. I would argue that you _should, when reasonable_, "expect" a chain to expose a part of its business logic from within its WASM blob (in the most un-opinionated form, a raw runtime-api[^1]), and then you build your pallet-api on top of it. 

For the sake of an example, let's say we expect open gov pallets to expose an api like:

```
fn votes_of(who: AccountId) -> Vec<VoteInfo>;
```

Needless to say, breaking this API is a red flag. So we can assume the API of this function will not change, and only implementation might. 

That simplifies some of the common data aggregation behind the scene for you.  

(I am not super familiar with this voting example, but I hope it makes sense)

This has 3 advantages: 

1. It removes a part of the maintenance cost from you, to polkadot-sdk developers. You'd still have to maintain your TS code, but you can 
2. It makes that piece of business logic inherently upgrade-able: no more nested `if version.is_low() { do this } else { do that }`. Once the WASM `:code` changes, the implementation changes automatically. 
3. This is quite SMOLDOT/light-client friendly. A light client can also consume this API and trust the answer that is coming out of it, through the virtue of trusting that the code stored under `:code` is correct, without needing to worry if a TS library is correct. 

I hope this gives a clear demonstration of why I advocate to:

1. Build a thin, but stable layer of runtime-level-APIs around common pallets
2. Then build more TS apis on top of it. 


[^1]: More complicated versions of it being pallet-view functions and XCQ, see [here](https://forum.polkadot.network/t/wasm-view-functions/1045).

**Comment by kianenigma** [comment]
But, regardless of what approach we take (Rust+TS or pure TS), we can call it a success as long as we hit one goal:

Once one person figures out the answer to a common query ("_how much is my f****ng tranferrable balance_?"), it should live somewhere that is reusable, and crucially, upgrade-able by all. 

Which means, no matter how this stuff is implemented, if the PAPI-runtime+pallet-api library becomes the standard, we can call it a success. 

On that end, looking in @IkerAula in here to see what he thinks.

**Comment by kianenigma** [comment]
> I acknowledge that this puts more computation onto the node though.

I think the approach you pointed out in the next paragraph is important and should be the norm: Technology-wise, a JS client has all the tools needed to have a wasm executor in its environment, and make queries to its own mini runtime, not going thought the `state_call` RPC. Indeed, if everyone starts doing this, it would be more pressure on nodes. 

This SMOLDOT-based approach IMO is the ultimate argument in favor of coding more things in runtime apis. 

Perhaps it would be great to expose the wasm-executor part of SMOLDOT and make it an individual lib. I am sure the code exists, it is just a matter of extracting it. 

@josepot @Tbaut do either of you know this? 

actually, what @xlc's chopsticks does it under the hood exactly the same. Probably there is some `chopsticks-core` on npm that would do this.

**Comment by carlosala** [comment]
We do have something like it in https://github.com/polkadot-api/wasm-executor. We only use it to get the metadata though, but it could definitely be extended. My concern regarding this would be the bundle sizes of it, it's a big wasm blob.

**Comment by kianenigma** [comment]
> We do have something like it in https://github.com/polkadot-api/wasm-executor. We only use it to get the metadata though, but it could definitely be extended. My concern regarding this would be the bundle sizes of it, it's a big wasm blob.

Do you have to download the wasm blob in the bundle though? It can probably be loaded in a somewhat if a lazy manner, so the application/page can start functioning, but to do runtime APIs you need to wait a few seconds for the wasm to be downloaded from the chain state.

**Comment by jsdw** [comment]
>  Indeed, if everyone starts doing this, it would be more pressure on nodes.

@kianenigma did you mean here, less pressure on nodes?

> Perhaps it would be great to expose the wasm-executor part of SMOLDOT and make it an individual lib. I am sure the code exists, it is just a matter of extracting it.

I became aware of it quite recently as in this PR we use substrate libs to download and execute the metadata runtime API eg here, which turns out not to be very many lines of code: https://github.com/paritytech/subxt/pull/1720/files#diff-a1681e673906d0317924847d770c99bd1079f6bbba3620f29357f950e37b1d00R39

Although I guess the trickier bit is also hooking up storage requests etc to acquire the relevant storage data for some calls (I guess what Smoldot is doing), but even without this, "pure" calls could exist that are super easy to run locally. These could expose things like the chains hashing logic, APIs to decode/provide info about some SCALE bytes that has custom encoding, etc.

Aside from Smoldot, I'm assuming try-runtime also does this sort of thing.

Either way, this is def something we could libraryise in Rust land in such a way that one could bind to in TS to benefit there too (though one may wish to re-implement parts of it in pure TS to avoid large wasm blob sizes).
